server/container/util/src/main/java/org/apache/james/util/ReactorUtils.java

```diff
@@ -38,25 +40,26 @@
 import reactor.core.publisher.Signal;
 import reactor.core.publisher.SynchronousSink;
 import reactor.util.context.Context;
-import reactor.util.function.Tuple2;
 
 public class ReactorUtils {
-
+    private static final Logger LOGGER = LoggerFactory.getLogger(ReactorUtils.class);
     public static final String MDC_KEY_PREFIX = "MDC-";
 
-    private static final Duration DELAY = Duration.ZERO;
-
     public static <T, U> RequiresQuantity<T, U> throttle() {
         return elements -> duration -> operation -> {
             Preconditions.checkArgument(elements > 0, "'windowMaxSize' must be strictly positive");
             Preconditions.checkArgument(!duration.isNegative(), "'windowDuration' must be strictly positive");
             Preconditions.checkArgument(!duration.isZero(), "'windowDuration' must be strictly positive");
 
             return flux -> flux
-                .windowTimeout(elements, duration)
-                .zipWith(Flux.interval(DELAY, duration))
-                .flatMap(Tuple2::getT1, elements, elements)
-                .flatMap(operation, elements);
+                .onErrorContinue((e, o) -> LOGGER.error("Error encountered while generating throttled entries", e))
+                .window(elements)
+                .delayElements(duration)
+                .concatMap(window -> window.flatMap(operation)
```

Comment:
I do not get how it work, can you explain it?

Reply:
We produce fixed windows, each one get delayed, and for each one we apply the transformation.

From the previous version, it resolves some reactor bugs. See tests for limitations.